source/slang/ir-sccp.cpp

```diff
@@ -0,0 +1,933 @@
+                // to see if we can narrow down which of the two ways
+                // might actually be taken.
+                //
+                auto condVal = getLatticeVal(conditionalBranch->getCondition());
+
+                // We do not expect to see a `None` value here, because that
+                // would mean the user is branching based on an undefined
+                // value.
+                //
+                // TODO: We should make sure there is no way for the user
+                // to trigger this assert with bad code that involves
+                // uninitialized variables. Right now we don't special
+                // case the `undefined` instruction when computing lattice
+                // values, so it shouldn't be a problem.
+                //
+                SLANG_ASSERT(condVal.flavor != LatticeVal::Flavor::None);
+
+                // If the branch condition is a constant, we expect it to
+                // be a Boolean constant. We won't assert that it is the
+                // case here, just to be defensive.
+                //
+                if( condVal.flavor == LatticeVal::Flavor::Constant )
+                {
+                    if( auto boolConst = as<IRBoolLit>(condVal.value) )
+                    {
+                        // Only one of the two targe blocks is possible to
+                        // execute, based on what we know of the condition,
+                        // so we will add that target to our work list and
+                        // bail out now.
+                        //
+                        auto target = boolConst->getValue() ? conditionalBranch->getTrueBlock() : conditionalBranch->getFalseBlock();
+                        cfgWorkList.Add(target);
+                        return;
+                    }
+                }
+
+                // As a fallback, if the condition isn't constant
+                // (or somehow wasn't a Boolean constnat), we will
+                // assume that either side of the branch could be
+                // taken, so that both of the target blocks are
+                // potentially executed.
+                //
+                cfgWorkList.Add(conditionalBranch->getTrueBlock());
+                cfgWorkList.Add(conditionalBranch->getFalseBlock());
+            }
+            else if( auto switchInst = as<IRSwitch>(inst) )
+            {
+                // The handling of a `switch` instruction is similar to the
+                // case for a two-way branch, with the main difference that
+                // we have to deal with an integer condition value.
+
+                auto condVal = getLatticeVal(switchInst->getCondition());
+                SLANG_ASSERT(condVal.flavor != LatticeVal::Flavor::None);
+
+                UInt caseCount = switchInst->getCaseCount();
+                if( condVal.flavor == LatticeVal::Flavor::Constant )
+                {
+                    if( auto condConst = as<IRIntLit>(condVal.value) )
```

Comment:
I suppose when we have constant folding - this will automatically cover more sophisticated scenarios.